pkg/netpol/eval/check.go

```diff
@@ -270,82 +270,163 @@ func (pe *PolicyEngine) allAllowedConnectionsBetweenPeers(srcPeer, dstPeer Peer)
+	// isCaptured : indicates if the xgress connections are captured by the policies in the current layer
+	isCaptured bool
+	// layerConns : the xgress connections between the peers / (peer and entire-cluster) from all relevant policies in the current layer
+	layerConns *k8s.PolicyConnections
+}
+
+// initEmptyPoliciesLayerXgressConns returns a new empty policiesLayerXgressConns object
+func initEmptyPoliciesLayerXgressConns() *policiesLayerXgressConns {
+	return &policiesLayerXgressConns{isCaptured: false, layerConns: k8s.NewPolicyConnections()}
+}
+
 // allAllowedXgressConnections: returns the allowed connections from srcPeer to dstPeer on the
 // given direction (ingress/egress)
 // allowed conns are computed by walking through all the available resources of k8s network policy api:
 // admin-network-policies, network-policies and baseline-admin-network-policies;
 // considering the precedence of each policy
+// in case of exposure-analysis it also checks and updates if a src is exposed to all namespaces on egress
+// or dst is exposed to all namespaces on ingress
 func (pe *PolicyEngine) allAllowedXgressConnections(src, dst k8s.Peer, isIngress bool) (allowedConns *common.ConnectionSet, err error) {
 	// first get allowed xgress conn between the src and dst from the ANPs
+	// (in case of exposure-analysis get also cluster wide conns of the selected peer from the ANPs)
 	// note that:
-	// - anpConns may contain allowed, denied or/and passed connections
-	// - anpCaptured is true iff there is at least one rule in the input ANPs that captures both src and dst;
+	// - anpConns.layerConns (or anpExposure.layerConns) may contain allowed, denied or/and passed connections;
+	// connections consider rules and policies precedence
+	// - anpConns.isCaptured is true iff there is at least one rule in the input ANPs that captures both src and dst;
 	// because anp rules are read as is and don't contain any implicit isolation effects for the Pods selected by the AdminNetworkPolicy.
-	anpConns, anpCaptured, err := pe.getAllAllowedXgressConnectionsFromANPs(src, dst, isIngress)
+	// anpExposure - is relevant only if pe.exposureAnalysisFlag == true
+	// - anpExposure.isCaptured is true iff there is at least one ANP that exposes the relevant peer (src on egress/ dst on ingress)
+	// to entire-cluster
+	anpConns, anpExposure, err := pe.getAllAllowedXgressConnectionsFromANPs(src, dst, isIngress)
 	if err != nil {
 		return nil, err
 	}
-	// optimization: if all the conns between src and dst were determined by the ANPs : return the allowed conns
-	if anpCaptured && anpConns.DeterminesAllConns() {
-		return anpConns.AllowedConns, nil
+	// optimizations
+	podExposureUpdatedFlag := false
+	anpConnsDeterminedAllFlag := false
+	// if all cluster-wide conns were determined by the ANP then update the selected peer's cluster wide exposure
+	if pe.exposureAnalysisFlag && anpExposure.layerConns.DeterminesAllConns() {
+		podExposureUpdatedFlag = true
```

Comment:
isn't exposure to external ip-blocks determined by regular netpols ?

Reply:
not only;
egress rules in (B)ANP can also determine conns to IPs (networks field)

Reply:
can `podExposureUpdatedFlag` be set to true, if possibly more exposure should be inferred from regular netpols?

Reply:
i think this is related to the other [comment](https://github.com/np-guard/netpol-analyzer/pull/486#discussion_r1930113088) the answer is yes

Reply:
it's not clear what do these flags mean?  where is it documented?

Reply:
consider removing "entire-world" fields from exposure analysis structs, and mention in documentation that exposure analysis only computes exposure to entire-cluster (for external ip-blocks just copying the conn-list results).

Reply:
opened #488